gpt_index/indices/query/base.py

```diff
@@ -235,28 +230,42 @@ def _get_nodes_for_response(
     ) -> List[Node]:
         """Get nodes for response."""
 
+    def _get_extra_info_for_response(
+        self,
+        nodes: List[Node],
+    ) -> Optional[Dict[str, Any]]:
+        """Get extra info for response."""
+        return None
+
     def _query(self, query_bundle: QueryBundle) -> Response:
         """Answer a query."""
+        self.response_builder.reset()
         # TODO: remove _query and just use query
         tuples = self.get_nodes_and_similarities_for_response(query_bundle)
-        node_texts = []
+
         for node, similarity in tuples:
             text, response = self._get_text_from_node(query_bundle, node)
-            self.response_builder.add_node(node, similarity=similarity)
+            self.response_builder.add_node_as_source(node, similarity=similarity)
             if response is not None:
                 # these are source nodes from within this node (when it's an index)
                 for source_node in response.source_nodes:
                     self.response_builder.add_source_node(source_node)
-            node_texts.append(text)
+            self.response_builder.add_text_chunks([text])
 
         if self._response_mode != ResponseMode.NO_TEXT:
-            response_str = self._give_response_for_nodes(
-                query_bundle.query_str, node_texts
-            )
+            response_str = self._give_response_for_nodes(query_bundle.query_str)
         else:
             response_str = None
 
-        return Response(response_str, source_nodes=self.response_builder.get_sources())
+        response_extra_info = self._get_extra_info_for_response(
```

Comment:
cc @logan-markewich i assume what you'd want is to actually parse the relationships as python tuples, not only as strings (to give you more flexibility). what i'm proposing in these changes are: 
1) Store kg_rel_texts (the relationship stringified text) and kg_rel_map (the python tuple of keyword to relationship) in the extra info of the response object (you can access via response.extra_info)
2) Define a `_get_extra_info_for_response` method to pass in these fields into the Response extra info
3) These are passed in from the node_info of each node (not the extra_info of each Node). It's a bit confusing, but stuff stored in the extra_info is injected into the node text, whereas stuff stored in the node_info is not.

Reply:
hope that works! i'm hoping to land this soon